src/qibolab/platforms/multiqubit.py

```diff
@@ -45,14 +139,44 @@ def execute_pulse_sequence(self, sequence: PulseSequence, nshots=None):
             if "readout" in roles[name]:
                 if not instrument_pulses[name].is_empty:
                     if not instrument_pulses[name].ro_pulses.is_empty:
+                        if all([pulse.serial in changed for pulse in instrument_pulses[name].ro_pulses]):
+                            # FIXME: for precision sweep in resonator spectroscopy
+                            # change necessary to perform precision sweep
+                            # TODO: move this to instruments (ask Alvaro)
+                            # TODO: check if this will work with multiplex
+                            for sequencers in self.instruments[name]._sequencers.values():
+                                for sequencer in sequencers:
+                                    sequencer.pulses = instrument_pulses[name].ro_pulses
                         results = self.instruments[name].acquire()
                         existing_keys = set(acquisition_results.keys()) & set(results.keys())
                         for key, value in results.items():
                             if key in existing_keys:
                                 acquisition_results[key].update(value)
                             else:
                                 acquisition_results[key] = value
-        return acquisition_results
+
+        # change back the frequency of the pulses
+        for name in self.instruments:
+            roles[name] = self.settings["instruments"][name]["roles"]
+            if "readout" in roles[name]:
+                instrument_pulses[name] = sequence.get_channel_pulses(*self.instruments[name].channels)
+                for pulse in instrument_pulses[name]:
+                    if pulse.serial in changed:
+                        pippo = acquisition_results[pulse.serial]
+                        # if abs(pulse.frequency) > 300e6:
+                        pulse.frequency += self.get_lo_readout_frequency(pulse.qubit)
+                        acquisition_results[pulse.serial] = pippo
+            if "control" in roles[name]:
+                instrument_pulses[name] = sequence.get_channel_pulses(*self.instruments[name].channels)
+                for pulse in instrument_pulses[name]:
+                    if pulse.serial in changed:
+                        pulse.frequency += self.get_lo_drive_frequency(pulse.qubit)
+
+        for key in ro_pulses:
+            data[key] = ExecutionResults(
+                acquisition_results[key][0], acquisition_results[key][1], np.array(acquisition_results[key][2])
+            )
```

Comment:
This is good. The only thing that we could add (which exists in the current main) is that apart from the pulse serials we could also have the qubit numbers as dictionary keys and map to the latest `ExecutionResults` for this qubit. This is useful for debugging and parsing the result when executing circuits.

Reply:
I've removed it since I thought that it was redundant. I can add it back. I was not so sure because I don't know if for some particular routines we will need to apply more than one readout pulse for the same qubit. If this is not the case we can use directly the qubit as a key instead of the serial of the pulse.